Dynamically recalculate the QuickPickItems

rmanthorpe opened the pull request:
The current implementations calls `git grep` once with no query and then
used the default `QuickPick` behaviour to filter the result. This update
dynamically updates the items as the user types. This starts to filter
and then the result list is updated with the grep results when they return.

rmanthorpe committed:
Dynamically recalculate the QuickPickItems

The current implementations calls `git grep` once with no quary and then
used the default `QuickPick` behaviour to filter the result. This update
dynamically updates the items as the user types. This starts to filter
and then the result list is updated with the grep results when they return.

rmanthorpe commented:
This is particularly important because at the moment if the empty `git grep` returns more data than fits in `exec`'s `maxBuffer` size then results are missed.

kawamurakazushi commented:
@rmanthorpe 
Thanks for your PR!

Actually first I tried this approach but it seems to be super slow, executing `git grep` on very value changed.
That's why I actually compromised with the current solution :( 
Though the `maxBuffer` issue is a critical bug..

Did you find any performance issue with your solutions?

rmanthorpe committed:
Kill unfinished greps

Calling `git grep` on every value change could be a performance issue.
This cancells unfinished calls to `git grep` when the next value change
occurs.

rmanthorpe commented:
I didn't run into any problems at all on a code base with ~200k SLOC. However, this solution isn't perfect - sometimes the execution order can be wrong. I've been experimenting with killing the previous git grep command if it's not completed when the value changes, which is slightly slower to load results (when you type very quickly) but ultimately a more robust solution I think.

rmanthorpe commented:
@kawamurakazushi I've added that change. This also now makes #5 redundant as it made sense to incorporate it here along with all the cancellation stuff.

kawamurakazushi commented:
Thanks! please give me some time to test it and review!

rmanthorpe commented:
Thanks! Let me know how performance is for you now.

rmanthorpe committed:
Merge branch 'master' into dynamic-items

rmanthorpe committed:
Use vscode.CancellationToken
